stl/inc/ranges

```diff
@@ -2962,8 +2961,16 @@ namespace ranges {
 
     template <class _Vw>
     class _Join_view_base : public view_interface<join_view<_Vw>> {
+    private:
+        struct _Cache_wrapper {
+            template <input_iterator _Iter>
+            constexpr _Cache_wrapper(const _Iter& _It) noexcept(noexcept(*_It)) : _Val(*_It) {}

```

Comment:
Is this guaranteed to be unambiguous when `_Iter` has a conversion function template (`template<class T> operator T();`)?

Not having to reason about this is why I used a tag type (though really any extra parameter should work)

Reply:
Urgh conversions of iterators

Reply:
I reused `_Not_quite_object::_Construct_tag` for that. @CaseyCarter shout if you want an explicit tag for that